extension/js/tree.js

```diff
@@ -88,22 +98,77 @@ function generateTestData() {
 
 function donateButtonClicked() {
 	console.log("donate clicked");
+	chrome.tabs.create({ url: "https://github.com/connorjoleary/DeepCite#contributions" });
 }
 
-function upvoteButtonClicked(element) {
+async function grab_ip() {
+	const response = await fetch('http://api.ipify.org/?format=json');
+    const data = await response.json();
+    return data.ip;
+}
+
+async function upvoteButtonClicked(event) {
+	var ipValue = await grab_ip();
+
+	var element = event.srcElement
 	var citeID = findClosestCiteID(element);
+
+	data = {type: "source",
+			ip: ipValue,
```

Comment:
Oof spacing, we should probably set up [prettierjs](https://prettier.io/). which I can do and also you can add it to git hooks so you don't have to remember to use it all the time!